### PR TITLE
8314438: NMT: Performance benchmarks are needed to have a baseline for comparison of improvements

### PR DESCRIPTION
**Pull request by afshin-zafari**
A new benchmark for  measuring the NMT overhead in `summary` and `detail` modes.
The tests are run using: 
```
make CONF=debug test TEST="micro:java.vm.runtime.NMTBenchmark" MICRO="RESULTS_FORMAT=json"
```
The results are written to a JSON file that can be visualized using [JMH Visualizer](https://jmh.morethan.io/).

### Notes
A separate [issue](https://bugs.openjdk.org/browse/JDK-8316814) is created for preparing a progfram for validating and analyzing the JMH outputs.
Another separate [issue](https://bugs.openjdk.org/browse/JDK-8316813) is created for measuring the virtual memory tracing parts of NMT.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8314438](https://bugs.openjdk.org/browse/JDK-8314438): NMT: Performance benchmarks are needed to have a baseline for comparison of improvements (**Enhancement** - P4)

### Reviewers
 * [Gerard Ziemski](https://openjdk.org/census#gziemski) (@gerard-ziemski - Committer) ⚠️ Review applies to [f6a153fc](https://git.openjdk.org/jdk/pull/15563/files/f6a153fcb5f93d4367faa31162dfbea0ef3adc47)
 * [Magnus Ihse Bursie](https://openjdk.org/census#ihse) (@magicus - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15563/head:pull/15563` \
`$ git checkout pull/15563`

Update a local copy of the PR: \
`$ git checkout pull/15563` \
`$ git pull https://git.openjdk.org/jdk.git pull/15563/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15563`

View PR using the GUI difftool: \
`$ git pr show -t 15563`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15563.diff">https://git.openjdk.org/jdk/pull/15563.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15563#issuecomment-1733084408)

### CODE REVIEW

**Review by gerard-ziemski — CHANGES_REQUESTED** [review]
The test will not compile for me unless I add:

`import java.util.concurrent.TimeUnit;`

**Comment by gerard-ziemski** [comment]
These are my results:

```
Benchmark                                   (N)  (THREADS)  (WITH_THREADS)  Mode  Cnt     Score    Error  Units
NMTBenchmark.NMTDetail.allocateMemory    100000          2               0  avgt   10   130.946 ?  8.761  ms/op
NMTBenchmark.NMTDetail.allocateMemory    100000          2               1  avgt   10     0.399 ?  0.228  ms/op
NMTBenchmark.NMTDetail.allocateMemory    100000          4               0  avgt   10   120.379 ?  8.072  ms/op
NMTBenchmark.NMTDetail.allocateMemory    100000          4               1  avgt   10     0.543 ?  0.207  ms/op
NMTBenchmark.NMTDetail.allocateMemory    100000          8               0  avgt   10   122.144 ?  3.696  ms/op
NMTBenchmark.NMTDetail.allocateMemory    100000          8               1  avgt   10     0.776 ?  0.132  ms/op
NMTBenchmark.NMTDetail.allocateMemory   1000000          2               0  avgt   10  1546.584 ? 24.937  ms/op
NMTBenchmark.NMTDetail.allocateMemory   1000000          2               1  avgt   10     2.757 ?  0.740  ms/op
NMTBenchmark.NMTDetail.allocateMemory   1000000          4               0  avgt   10  1530.738 ? 73.529  ms/op
NMTBenchmark.NMTDetail.allocateMemory   1000000          4               1  avgt   10     2.859 ?  0.624  ms/op
NMTBenchmark.NMTDetail.allocateMemory   1000000          8               0  avgt   10  1609.203 ? 29.717  ms/op
NMTBenchmark.NMTDetail.allocateMemory   1000000          8               1  avgt   10     3.113 ?  0.632  ms/op
NMTBenchmark.NMTOff.allocateMemory       100000          2               0  avgt   10    60.585 ?  4.187  ms/op
NMTBenchmark.NMTOff.allocateMemory       100000          2               1  avgt   10     0.333 ?  0.149  ms/op
NMTBenchmark.NMTOff.allocateMemory       100000          4               0  avgt   10    56.138 ?  2.727  ms/op
NMTBenchmark.NMTOff.allocateMemory       100000          4               1  avgt   10     0.493 ?  0.214  ms/op
NMTBenchmark.NMTOff.allocateMemory       100000          8               0  avgt   10    57.457 ?  2.310  ms/op
NMTBenchmark.NMTOff.allocateMemory       100000          8               1  avgt   10     0.731 ?  0.158  ms/op
NMTBenchmark.NMTOff.allocateMemory      1000000          2               0  avgt   10   815.515 ? 19.232  ms/op
NMTBenchmark.NMTOff.allocateMemory      1000000          2               1  avgt   10     2.674 ?  0.566  ms/op
NMTBenchmark.NMTOff.allocateMemory      1000000          4               0  avgt   10   829.726 ? 31.153  ms/op
NMTBenchmark.NMTOff.allocateMemory      1000000          4               1  avgt   10     2.678 ?  0.652  ms/op
NMTBenchmark.NMTOff.allocateMemory      1000000          8               0  avgt   10   817.929 ? 19.142  ms/op
NMTBenchmark.NMTOff.allocateMemory      1000000          8               1  avgt   10     3.023 ?  0.578  ms/op
NMTBenchmark.NMTSummary.allocateMemory   100000          2               0  avgt   10   101.419 ?  4.231  ms/op
NMTBenchmark.NMTSummary.allocateMemory   100000          2               1  avgt   10     0.329 ?  0.143  ms/op
NMTBenchmark.NMTSummary.allocateMemory   100000          4               0  avgt   10   102.224 ?  3.317  ms/op
NMTBenchmark.NMTSummary.allocateMemory   100000          4               1  avgt   10     0.514 ?  0.230  ms/op
NMTBenchmark.NMTSummary.allocateMemory   100000          8               0  avgt   10   101.127 ?  4.144  ms/op
NMTBenchmark.NMTSummary.allocateMemory   100000          8               1  avgt   10     0.719 ?  0.168  ms/op
NMTBenchmark.NMTSummary.allocateMemory  1000000          2               0  avgt   10  1423.287 ? 56.109  ms/op
NMTBenchmark.NMTSummary.allocateMemory  1000000          2               1  avgt   10     2.682 ?  0.442  ms/op
NMTBenchmark.NMTSummary.allocateMemory  1000000          4               0  avgt   10  1418.999 ? 33.891  ms/op
NMTBenchmark.NMTSummary.allocateMemory  1000000          4               1  avgt   10     2.812 ?  0.510  ms/op
NMTBenchmark.NMTSummary.allocateMemory  1000000          8               0  avgt   10  1431.552 ? 29.915  ms/op
NMTBenchmark.NMTSummary.allocateMemory  1000000          8               1  avgt   10     3.073 ?  0.583  ms/op
```

You said `The JSON file can be used for visualising the results.` Can you please explain how to do that exactly?

**Comment by gerard-ziemski** [comment]
It looks like a good start.

**Comment by gerard-ziemski** [comment]
The test took something like 20 minutes for me.

Is the multithreading mode really useful?

And even if we decide that it is, we can probably skip the entries that create the threads, but not actually use them? Ex:

THREADS=2, WITH_THREADS=0
THREADS=4, WITH_THREADS=0
THREADS=8, WITH_THREADS=0

**Comment by afshin-zafari** [comment]
> The test will not compile for me unless I add:
> 
> `import java.util.concurrent.TimeUnit;`

Sorry, it was removed mistakenly before pushing the code. Now fixed.

**Comment by afshin-zafari** [comment]
> You said `The JSON file can be used for visualising the results.` Can you please explain how to do that exactly?

The [visualizer site](https://jmh.morethan.io) is added to the PR description. The JSON file can be dropped there.

**Comment by afshin-zafari** [comment]
> Is the multithreading mode really useful?

NMT records stack traces for threads in a different way as of other threads. It uses a link list for the former and a static hash table for the latter. I tried to make a more realistic use case of NMT.

**Comment by afshin-zafari** [comment]
> THREADS=2, WITH_THREADS=0
> THREADS=4, WITH_THREADS=0
> THREADS=8, WITH_THREADS=0

The last two rows of the above list are skipped now. Thanks for the comment.

**Comment by gerard-ziemski** [comment]
I like it! I do have more follow ups, tough.

I see that we are currently mixing malloc/free. Wouldn't we better served if we separated those as 2 distinct categories? I can imagine a scenario where either malloc or free path could be regressed/improved without affecting the other.

Also, I think we should add realloc to the mix, again keeping it separate. I tried to collapse malloc/realloc into single code path, but it was rejected, mostly on the basis that it would impact the performance - this microbenchmark would be a perfect tool to offer a definitive answer.

**Comment by afshin-zafari** [comment]
> The tool should look at `NMTOff` first from both runs and verify that they are the same, to prove that the results are useful (there should be no difference)

Which measures should be the same between *both runs*? For `NMT_Off` for example, there are 2 (no of threads= 0,4) measures per method. Consider also, we will have more methods for virtual memory alloc/dealloc tests.

**Comment by gerard-ziemski** [comment]
> > The tool should look at `NMTOff` first from both runs and verify that they are the same, to prove that the results are useful (there should be no difference)
> 
> Which measures should be the same between _both runs_? For `NMT_Off` for example, there are 2 (no of threads= 0,4) measures per method. Consider also, we will have more methods for virtual memory alloc/dealloc tests.

I think all the available data should be used to verify, so all runs with `NMT_Off`?

This can be done in a followup issue I think?

To me it looks like we have a good starting point, which we can use already (by manually comparing the results), so we are OK to push this in I think.

**Review by gerard-ziemski — APPROVED** [review]

File: test/micro/org/openjdk/bench/java/util/NMTBenchmark.java [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package org.openjdk.bench.java.util;
```
Comment (AlanBateman) [inline test/micro/org/openjdk/bench/java/util/NMTBenchmark.java:24]: The micros in test/micro/org/openjdk/bench/java/util are usually to test methods in java.util.**. Should this be moved to test/micro/org/openjdk/bench/vm/runtime?

Comment (afshin-zafari) [reply test/micro/org/openjdk/bench/java/util/NMTBenchmark.java]: Thanks for the comment. 
Fixed now.

**Review by magicus — APPROVED** [review]
Build changed trivially good.

Should there not be a label to alert NMT folks for reviewing?

**Comment by afshin-zafari** [comment]
> Should there not be a label to alert NMT folks for reviewing?
Thanks for the comment. 

/cc hotspot-runtime

**Pull request closed without merging**